### ISSUE TITLE
Unable to install on python3.10

### ISSUE THREAD

**Issue by academe-01** [h5py/h5py#2251]
Any help much appreciated !

HDF5 Version: (1, 10, 10)
Python: 3.10
mpi4py: 3.1.1

```
root@x:/tmp/1/h5py# CC=mpicc HDF5_MPI="ON" python3.10 -m pip install --no-binary=h5py h5py
DEPRECATION: --no-binary currently disables reading from the cache of locally built wheels. In the future --no-binary will not influence the wheel cache. pip 23.1 will enforce this behaviour change. A possible replacement is to use the --no-cache-dir option. You can use the flag --use-feature=no-binary-enable-wheel-cache to test the upcoming behaviour. Discussion can be found at https://github.com/pypa/pip/issues/11453
Collecting h5py
  Using cached h5py-3.8.0.tar.gz (400 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... error
  error: subprocess-exited-with-error
  
  × pip subprocess to install backend dependencies did not run successfully.
  │ exit code: 1
  ╰─> [23 lines of output]
      DEPRECATION: --no-binary currently disables reading from the cache of locally built wheels. In the future --no-binary will not influence the wheel cache. pip 23.1 will enforce this behaviour change. A possible replacement is to use the --no-cache-dir option. You can use the flag --use-feature=no-binary-enable-wheel-cache to test the upcoming behaviour. Discussion can be found at https://github.com/pypa/pip/issues/11453
      Collecting mpi4py==3.1.0
        Using cached mpi4py-3.1.0.tar.gz (2.4 MB)
        Installing build dependencies: started
        Installing build dependencies: finished with status 'done'
        Getting requirements to build wheel: started
        Getting requirements to build wheel: finished with status 'error'
        error: subprocess-exited-with-error
      
        × Getting requirements to build wheel did not run successfully.
        │ exit code: 1
        ╰─> [1 lines of output]
            error in mpi4py setup command: 'python_requires' must be a string containing valid version specifiers; Invalid specifier: '!=3.4.*''
            [end of output]
      
        note: This error originates from a subprocess, and is likely not a problem with pip.
      error: subprocess-exited-with-error
      
      × Getting requirements to build wheel did not run successfully.
      │ exit code: 1
      ╰─> See above for output.
      
      note: This error originates from a subprocess, and is likely not a problem with pip.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× pip subprocess to install backend dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```

build against h5py from git
```
root@x:/tmp/1/h5py# CC=mpicc HDF5_MPI="ON" python3.10 setup.py build
/root/.local/lib/python3.10/site-packages/setuptools/config/pyprojecttoml.py:102: _ExperimentalProjectMetadata: Support for project metadata in `pyproject.toml` is still experimental and may be removed (or change) in future releases.
  warnings.warn(msg, _ExperimentalProjectMetadata)
running build
running build_py
running build_ext
Loading library to get build settings and version: libhdf5.so
********************************************************************************
                       Summary of the h5py configuration

  HDF5 include dirs: []
  HDF5 library dirs: []
       HDF5 Version: (1, 10, 10)
        MPI Enabled: True
   ROS3 VFD Enabled: True
 DIRECT VFD Enabled: True
   Rebuild Required: True
     MS-MPI Enabled: False
MS-MPI include dirs: []
MS-MPI library dirs: []

********************************************************************************
Executing cythonize()
[ 1/24] Cythonizing /tmp/1/h5py/h5py/_errors.pyx

Error compiling Cython file:
------------------------------------------------------------
...

cdef struct err_data_t:
    H5E_error_t err
    int n

cdef herr_t walk_cb(unsigned int n, const H5E_error_t *desc, void *e) nogil noexcept:
                                                                           ^
------------------------------------------------------------

h5py/_errors.pyx:97:76: Syntax error in C variable declaration
Traceback (most recent call last):
  File "/tmp/1/h5py/setup.py", line 68, in <module>
    setup(
  File "/root/.local/lib/python3.10/site-packages/setuptools/__init__.py", line 87, in setup
    return distutils.core.setup(**attrs)
  File "/root/.local/lib/python3.10/site-packages/setuptools/_distutils/core.py", line 148, in setup
    return run_commands(dist)
  File "/root/.local/lib/python3.10/site-packages/setuptools/_distutils/core.py", line 163, in run_commands
    dist.run_commands()
  File "/root/.local/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 967, in run_commands
    self.run_command(cmd)
  File "/root/.local/lib/python3.10/site-packages/setuptools/dist.py", line 1214, in run_command
    super().run_command(command)
  File "/root/.local/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 986, in run_command
    cmd_obj.run()
  File "/root/.local/lib/python3.10/site-packages/setuptools/_distutils/command/build.py", line 136, in run
    self.run_command(cmd_name)
  File "/root/.local/lib/python3.10/site-packages/setuptools/_distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/root/.local/lib/python3.10/site-packages/setuptools/dist.py", line 1214, in run_command
    super().run_command(command)
  File "/root/.local/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 986, in run_command
    cmd_obj.run()
  File "/tmp/1/h5py/setup_build.py", line 171, in run
    self.extensions = cythonize(self._make_extensions(config),
  File "/usr/local/lib/python3.10/site-packages/Cython/Build/Dependencies.py", line 1127, in cythonize
    cythonize_one(*args)
  File "/usr/local/lib/python3.10/site-packages/Cython/Build/Dependencies.py", line 1250, in cythonize_one
    raise CompileError(None, pyx_file)
Cython.Compiler.Errors.CompileError: /tmp/1/h5py/h5py/_errors.pyx
```

**Comment by academe-01** [comment]
the same on python3.8

**Comment by tacaswell** [comment]
These issue are not related to the version of Python.   The first issue is a duplicate of https://github.com/h5py/h5py/issues/2222 which is caused by `setuptools` becoming stricter about version specifiers and the build system hard-pinning the mpi4py version (which makes sense for building redistributable wheels but less sense when you are trying to just build for your self).


The second is due to the main branch needing a newer version of cython and by directly invoking `setup.py` you are side-stepping where we pinned that.


pre https://github.com/h5py/h5py/issues/2222#issuecomment-1408346741

> In terms of a workaround, you could install a working version of mpi4py and the other build dependencies beforehand, and then use pip's --no-build-isolation flag to build h5py with those rather than creating a new build env.

and making sure you have the correct build dependencies in your environment is probably the best path forward.

**Comment by yousefmoazzam** [comment]
Hi, I also had the same error as the first of the two in the original post: I had a conda environment with python 3.10.x in it where `pip` was failing to install `mpi4py` as a dependency of `h5py`. The TL;DR is that I managed to fix it by forcing it to install an `mpi4py` version greater than `3.1.0` via:
```
CC=$(type -p mpicc) HDF5_MPI="ON" HDF5_DIR=/path/to/hdf5 pip3 install --no-binary=h5py h5py mpi4py>3.1.0
```

The explanation for why I think this worked is a bit long, but if anyone interested, here it is...

I think that the error being given:
```
'python_requires' must be a string containing valid version specifiers; Invalid specifier: '!=3.4.*''
```
is what has been fixed in this `mpi4py` commit to mpi4py's `setup.py` file: https://github.com/mpi4py/mpi4py/commit/e77486bea7a7520d09cea48cd03a98d583a85f60.

Notice the date of that commit/fix: August **14**th 2021. The release of mpi4py v3.1.0 was on [August **12**th 2021](https://github.com/mpi4py/mpi4py/releases/tag/3.1.0), and therefore, the [v3.1.0 package on PyPI](https://pypi.org/project/mpi4py/3.1.0/) contains code from August 12th 2021. This is before the fix to the `setup.py` file.

The original one-liner I was running to install `h5py` with parallel hdf5 support (so `mpi4py` was a dependency) was:
```
CC=$(type -p mpicc) HDF5_MPI="ON" HDF5_DIR=/path/to/hdf5 pip3 install --no-binary=h5py h5py
```
and in particular, this was resolving the dependencies to get mpi4py **v3.1.0**; ie, the version with the typo in `setup.py`.

However, if I force it to install a version of `mpi4py` greater than 3.1.0 like so:
```
CC=$(type -p mpicc) HDF5_MPI="ON" HDF5_DIR=/path/to/hdf5 pip3 install --no-binary=h5py h5py mpi4py>3.1.0
```
then the error does not occur, and it installs `mpi4py` alongside `h5py` just fine (it gets mpi4py v3.1.4).

It's not definitive proof, but it was enough to make me suspicious that maybe the mpi4py 3.1.0 package on PyPI is broken (at least, in some cases) due to it not containing that fix to `setup.py`?...

Anyway, I hope that can be of some use with fixing that particular error, good luck! :slightly_smiling_face:
